packages/dds/tree/src/feature-libraries/indexing/anchorTreeIndex.ts

```diff
@@ -0,0 +1,285 @@
+ * but returns the cursor to the state it was in before being passed to the function. It should also not be disposed by this function
+ * and must be disposed elsewhere.
+ */
+export type KeyFinder<TKey extends TreeValue> = (tree: ITreeSubscriptionCursor) => TKey;
+
+/**
+ * An index from some arbitrary keys to anchor nodes. Keys can be anything that is a {@link TreeValue}.
+ * A key can map to multiple nodes but each collection of nodes only results in a single value.
+ *
+ * @remarks
+ * Detached nodes are stored in the index but filtered out when any public facing apis are called. This means that
+ * calling {@link keys} will not include any keys that are stored in the index but only map to detached nodes.
+ *
+ * TODO: need to make sure key finders are deterministic or have a way to invalidate them
+ */
+export class AnchorTreeIndex<TKey extends TreeValue, TValue>
+	implements TreeIndex<TKey, TValue>
+{
+	/**
+	 * Caches {@link KeyFinder}s for each schema definition. If a schema maps to null, it does not
+	 * need to be considered at all for this index. This allows us to skip subtrees that aren't relevant
+	 * as a performance optimization.
+	 */
+	private readonly keyFinders = new Map<TreeNodeSchemaIdentifier, KeyFinder<TKey> | null>();
+	/**
+	 * The actual index from keys to anchor nodes.
+	 */
+	private readonly nodes = new Map<TKey, TreeIndexNodes<AnchorNode>>();
+	/**
+	 * Keeps track of anchors for disposal.
+	 */
+	private readonly anchors = new Map<AnchorNode, Anchor>();
+
+	/**
+	 * @param forest - the forest that is being indexed
+	 * @param indexer - a function that retrieves the key finder based on a given schema or undefined if the schema does not have an associated key finder
+	 * @param getValue - a function that returns the value or undefined given at least one anchor node
+	 * @param checkTreeStatus - a function that gets the tree status from an anchor node, used for filtering out detached nodes
+	 */
+	public constructor(
+		private readonly forest: IForestSubscription,
+		indexer: (schemaId: TreeNodeSchemaIdentifier) => KeyFinder<TKey> | undefined,
+		private readonly getValue: (anchorNodes: TreeIndexNodes<AnchorNode>) => TValue | undefined,
+		private readonly checkTreeStatus: (node: AnchorNode) => TreeStatus | undefined,
+	) {
+		/**
+		 * Given a cursor in field mode, recursively indexes all nodes under the field.
+		 */
+		const indexField = (fieldCursor: ITreeSubscriptionCursor): void => {
+			forEachNode(fieldCursor, (nodeCursor) => {
+				const keyFinder = getOrCreate(
+					this.keyFinders,
+					fieldCursor.type,
+					// if the indexer does not return a key finder for this schema, we cache a null value to indicate the indexer
+					// does not need to be called if this schema is encountered in the future
+					(schema) => indexer(schema) ?? null,
+				);
+
+				if (keyFinder !== null) {
+					const key = keyFinder(fieldCursor);
```

Comment:
I'm generally confused why we are using fieldCursor and not nodeCursor. what am I missing?

Reply:
we're using field cursor because the schema definition is for the field. it also starts with the root field and we recursively walk the entire tree, using both field cursors and node cursors to index the entire tree